client/state/timezones/test/reducer.js

```diff
@@ -0,0 +1,335 @@
+/**
+ * External dependencies
+ */
+import { expect } from 'chai';
+import deepFreeze from 'deep-freeze';
+
+/**
+ * Internal dependencies
+ */
+import { useSandbox } from 'test/helpers/use-sinon';
+
+import timezonesReducer, {
+	byContinentsReducer,
+	labelsReducer,
+	rawOffsetReducer,
+} from '../reducer';
+
+import { timezonesReceive } from '../actions';
+import { fromApi } from 'state/data-layer/wpcom/timezones';
+
+/**
+ * Fixture data
+ */
+import {
+	RAW_OFFSETS,
+	TIMEZONES,
+	TIMEZONES_BY_CONTINENT,
+} from './fixture';
+
+describe( 'reducer', () => {
+	let sandbox;
+
+	useSandbox( newSandbox => {
+		sandbox = newSandbox;
+		sandbox.stub( console, 'warn' );
+	} );
+
+	it( 'should export expected reducer keys', () => {
+		expect( timezonesReducer( undefined, {} ) ).to.have.keys( [
+			'byContinents',
+			'labels',
+			'rawOffsets',
+		] );
+	} );
+
+	describe( '#rawOffsetReducer()', () => {
+		it( 'should default to an empty Object', () => {
+			expect( rawOffsetReducer( undefined, {} ) ).to.eql( {} );
+		} );
+
+		it( 'should index timezones.rawOffsets state', () => {
+			const initialState = undefined;
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should override timezones.rawOffsets state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: [
+					{ value: 'foo', label: 'bar' }
+				],
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} ).rawOffsets;
+
+			deepFreeze( initialState );
+
+			const restAPIResponse = {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			};
+
+			const apiResponse = fromApi( restAPIResponse );  // data-layer processing
+			const action = timezonesReceive( apiResponse );
+
+			const expectedState = apiResponse.rawOffsets;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
+
+		it( 'should persist state', () => {
+			const initialState = fromApi( {
+				manual_utc_offsets: RAW_OFFSETS,
+				timezones: TIMEZONES,
+				timezones_by_continent: TIMEZONES_BY_CONTINENT,
+			} );
+			deepFreeze( initialState );
+
+			const action = { type: 'SERIALIZE' };
+
+			const expectedState = initialState;
+			const newState = rawOffsetReducer( initialState, action );
+			expect( newState ).to.eql( expectedState );
+		} );
```

Comment:
> We're testing that we wrote down the schema we wanted, that the schema actually represents the data contract we want to have

exactly :smile: we could also consider creating some _property-based_ testing which would use the JSON schema to generate random test data. at some point we'll always hit the fact that we're manually encoding some structure. I'm not sure where the line is on testing that structure. I'm pretty sure it's drawn before we introduce 500 tests to make sure that the schema persists for various sub-stress

> We have a selenium test runner that performs a number of smoke tests at different viewport sizes for common flows. The tests currently run during every master branch merge.

This is cool. I would like to learn more about this and maybe communicate more about it generally.